Make it possible for a repository to multiple downloads in parallel

lberki opened the pull request.

lberki commented:
WDYT?

This is intended to be a quick and dirty attempt to fix #19674. There are a number of, erm, questionable points:

* I added Yet Another Threadpool. This is especially hairy because `HttpDownloader` already has some concept of parallelism, but works with a semaphore (see `MAX_PARALLEL_DOWNLOADS`). With Loom impending, I don't think it makes a lot of sense to invest a lot into implementing this in a more async way, but it would be nice if Bazel had a consistent view of how parallelism should work. Especially interesting is the thought process by which I came up with "16" as the number of max. threads : _"I made this pull request n like 2 hours and I didn't consider anything."_
* The reason why I added the `defer=` argument to `repository_ctx.download()` is that I was too lazy to copy-paste the big chunk of parameter type definitions into another method or to write my own type checking boilerplate code. I originally wanted to do something like `download_multiple([{"url": "http://foo.bar", "output": "foo"}, {"url": "http://bar.foo", "output": "bar"}])`, but that turned out to require a lot of hand-written type conversion / type checking code and would have emitted error messages inconsistent with the rest of Bazel anyway. The download actually starts on the `download_multiple()` call, which is a nice way to avoid dealing with a lot of hairiness like a Starlark function call starting a download then getting a SIGINT or a Starlark error before it can finish. 
* I'm not sure if it makes sense to set each parameter to a different value on each request. It's just what falls out from the current `download(defer=True); download_multiple()` API. You tell me if this is a nice API or not. I don't think so, but I haven't been able to come up with a better one this quickly.
* There is probably a nicer way of waiting for a list of futures to complete, but this is the only approach that I could come up with that works with a vanilla `Future` (as opposed to `CompletableFuture`, etc.)

Wyverald commented on src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkBaseExternalContext.java:
```diff
@@ -445,72 +579,49 @@ public StructImpl download(
+    Sequence<DownloadRequest> requests =
+        Sequence.cast(requestsUnchecked, DownloadRequest.class, "requests");
+
+    List<PendingDownload> downloads = new ArrayList<>();
```
could we merge `DownloadRequest` and `PendingDownload`? Then we only need a single class that essentially just wraps a `Future`. That does mean that the download starts as soon as you call `rctx.download(defer=True)`, but maybe that's a good thing.

Also at that point we could just let people wait on individual deferred downloads? That removes the need to return the awkward list of returned structs. `rctx.download(defer=True).execute()` can return what you expect.

lberki replied:
I thought the `rctx.download(defer=True).execute()` approach would make it impossible to do multiple downloads at the same time?

In the current implementation, the download doesn't actually start until the `download_multiple()` call, which is handy because it makes state management a bit easier because then it's certain that there are no HTTP downloads pending, except during the `download_multiple()` call. Do you think we should break that invariant and start the download immediately after the `download()` call and make things more complicated or do I misunderstand what you suggest above?

(then you'd be advocating for the `token` approach I mentioned here: https://github.com/bazelbuild/bazel/issues/19674#issuecomment-1753046186)

lberki replied:
My line of reasoning was that we should absolutely, positively, under no circumstances should leave a download running that was initiated by a repository rule if the processing of that repository rule terminates and this invariant seemed to be much easier to uphold if we also have the invariant "we only ever have downloads running when some sort of downloading function is on the call stack".

I'll take a look how dangerous it looks like to do it with the token approach as time allows (although now that Thanksgiving is over, my time is much more dear)

lberki replied:
Actually, it ended up being much more pleasant than I thought it would be.

Wyverald commented on src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/DownloadManager.java:
```diff
@@ -114,7 +177,7 @@ public void setCredentialFactory(CredentialFactory credentialFactory) {
    * @throws InterruptedException if this thread is being cast into oblivion
    */
-  public Path download(
+  public Path reallyDownload(
```
this can be private now, right?

lberki replied:
Done

lberki commented:
PTAL! I ended up doing the `token` approach as per your suggestion. The current API is:

```
request = rctx.download("https://foo.bar", output="foo", defer=True)
request.wait()
```

And I *think* ensuring the property "no work is going on after the repository function finishes executing, one way or the other" is implemented both correctly and simply enough to be "obviously" correct.

I do realize that test cases are completely missing, but first let's figure out if the implementation is roughly correct. I'm pretty sure there will be bugs uncovered by the test cases anyway.

fmeum commented:
I find it difficult to guess what the `defer` parameter does just based on its name: Does it defer the download until I start it later or does it start the download right away and instead defers returning the result (the latter, as far as I understand).

I could see `block` or `wait`, both with a default of `true`, being more descriptive.

lberki commented:
I'll block^Udefer to @Wyverald on that one; I haven't done a lot of careful consideration of various options, just put the first word in that came to my mind.

lberki commented:
I added test cases, made this into a non-draft since it's getting close to production quality and added an error message for pending asynchronous work at the end of repository rules. 

I think if you are fine with the current token-based implementation, the only big question that remains is whether the additional thread pool is okay and if not, what else to do. And the naming of the new method, of course, but I'm sure we can agree about that :)

lberki commented:
Thanks! I'l pull this into an internal change then and try to fix all the nits I think you might have (it's a little bit less toil to do it like that and I don't feel too bad about doing the nitpicking rounds internally)

lberki commented:
I just realized that using the identifiers `async` and `await` is probably a bad idea since those are keywords in Python. So I'm going with `block` / `wait` instead.

Wyverald commented:
This was merged as https://github.com/bazelbuild/bazel/commit/73c1a1e382cdd99766e4e11bb917e9672c11b5f3
